File: sonar-plugin-api/src/main/java/org/sonar/api/config/License.java
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.api.config;

import java.io.IOException;
import java.io.StringReader;
import java.nio.charset.StandardCharsets;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;
import org.apache.commons.codec.binary.Base64;
import org.apache.commons.io.IOUtils;
import org.apache.commons.lang.StringUtils;
import org.sonar.api.utils.DateUtils;

/**
 * SonarSource license. This class aims to extract metadata but not to validate or - of course -
 * to generate license
 *
 * @since 3.0
 */
public final class License {
  private String product;
  private String organization;
  private String expirationDate;
  private String type;
  private String server;
  private Map<String, String> additionalProperties;

  private License(Map<String, String> properties) {
    this.additionalProperties = new HashMap<>(properties);
    product = StringUtils.defaultString(get("Product", properties), get("Plugin", properties));
    organization = StringUtils.defaultString(get("Organisation", properties), get("Name", properties));
    expirationDate = StringUtils.defaultString(get("Expiration", properties), get("Expires", properties));
    type = get("Type", properties);
    server = get("Server", properties);
    // SONAR-4340 Don't expose Digest and Obeo properties
    additionalProperties.remove("Digest");
    additionalProperties.remove("Obeo");
  }

  private String get(String key, Map<String, String> properties) {
    additionalProperties.remove(key);
    return properties.get(key);
  }

  /**
   * Get additional properties available on this license (like threshold conditions)
   * @since 3.6
   */
  public Map<String, String> additionalProperties() {
    return additionalProperties;
  }

  @Nullable
  public String getProduct() {
    return product;
  }

  @Nullable
  public String getOrganization() {
    return organization;
  }

  @Nullable
  public String getExpirationDateAsString() {
    return expirationDate;
  }

  @Nullable
  public Date getExpirationDate() {
    return DateUtils.parseDateQuietly(expirationDate);
  }

  public boolean isExpired() {
    return isExpired(new Date());
  }

  boolean isExpired(Date now) {
    Date date = getExpirationDate();
    if (date == null) {
      return false;
    }
    // SONAR-6079 include last day
    Calendar cal = Calendar.getInstance();
    cal.setTime(date);
    cal.add(Calendar.DAY_OF_MONTH, 1);
    cal.add(Calendar.SECOND, -1);
    return now.after(cal.getTime());
  }

  @Nullable
  public String getType() {
    return type;
  }

  @Nullable
  public String getServer() {
    return server;
  }

  public static License readBase64(String base64) {
    return readPlainText(new String(Base64.decodeBase64(base64.getBytes(StandardCharsets.UTF_8)), StandardCharsets.UTF_8));
  }

  static License readPlainText(String data) {
    Map<String, String> props = new HashMap<>();
    try {
      List<String> lines = IOUtils.readLines(new StringReader(data));
      for (String line : lines) {
        if (StringUtils.isNotBlank(line) && line.indexOf(':') > 0) {
          String key = StringUtils.substringBefore(line, ":");
          String value = StringUtils.substringAfter(line, ":");
          props.put(StringUtils.trimToEmpty(key), StringUtils.trimToEmpty(value));
        }
      }

    } catch (IOException e) {
      // silently ignore
    }
    return new License(props);
  }
}

File: server/sonar-web/config/license.ts
/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { Dependency } from 'rollup-plugin-license';

// To be always synced with https://saas-eu.whitesourcesoftware.com/Wss/WSS.html#!policyDetails;policy=3131;org=318388
// Below is the list of approved front-end licenses extracted from Mend
export const ALLOWED_LICENSES = [
  '(MPL-2.0 OR Apache-2.0)', // Multiple licenses. Added specifically for dompurify as we have ignored this in Mend
  '0BSD',
  'Apache-2.0',
  'BSD-2-Clause',
  'BSD-3-Clause',
  'ISC',
  'LGPL-3.0',
  'MIT',
];

// Just for Sprig currently, it has an Apache-2 license that isn't correctly parsed by the plugin
export const ALLOWED_LICENSE_TEXT = ['http://www.apache.org/licenses/LICENSE-2.0'];

// Generates license information with necessary details.
// A package which has a valid license that the plugin is unable read will default to MIT
export const generateLicenseText = (dependency: Dependency) => {
  const { author, homepage, license, licenseText, name, repository, version } = dependency;
  const lines: string[] = [];

  lines.push(`Name: ${name}`);
  lines.push(`Version: ${version}`);

  if (license) {
    lines.push(`License: ${license}`);
  }

  if (typeof repository === 'string') {
    lines.push(`Repository: ${repository}`);
  } else if (repository?.url) {
    lines.push(`Repository: ${repository.url}`);
  } else if (homepage) {
    lines.push(`Homepage: ${homepage}`);
  }

  if (author) {
    lines.push(`Author: ${author.text()}`);
  }

  if (licenseText) {
    lines.push(`License Copyright:`);
    lines.push(`${licenseText}`);
  }

  return lines.join('\n');
};
